packages/uploadthing/src/client-future.ts

```diff
@@ -0,0 +1,402 @@
+  return uploadingFile;
+}
+
+/**
+ * Modifies an uploading file to an uploaded file in place
+ * @internal
+ */
+function transitionToUploaded<TRoute extends AnyFileRoute>(
+  file: UploadingFile,
+  xhrResult: UploadPutResult,
+): UploadedFile<TRoute> {
+  const uploadedFile = file as unknown as UploadedFile<TRoute>;
+  uploadedFile.status = "uploaded";
+  uploadedFile.data = xhrResult.serverData;
+  uploadedFile.hash = xhrResult.fileHash;
+  uploadedFile.url = xhrResult.ufsUrl;
+  return uploadedFile;
+}
+
+/**
+ * Modifies a pending or uploading file to a failed file in place
+ * @internal
+ */
+function transitionToFailed<TRoute extends AnyFileRoute>(
+  file: PendingFile | UploadingFile,
+  reason: UploadThingClientError<TRoute["$types"]["errorShape"]>,
+): FailedFile<TRoute> {
+  const failedFile = file as unknown as FailedFile<TRoute>;
+  failedFile.status = "failed";
+  failedFile.reason = reason;
+  return failedFile;
+}
+
+export interface PresignedReceivedEvent<TRoute extends AnyFileRoute> {
+  type: "presigned-received";
+  /**
+   * All files that are being uploaded as part of this action.
+   */
+  files: AnyFile<TRoute>[];
+}
+
+export interface UploadStartedEvent<TRoute extends AnyFileRoute> {
+  type: "upload-started";
+  /**
+   * The file that started uploading.
+   */
+  file: UploadingFile;
+  /**
+   * All files that are being uploaded as part of this action.
+   */
+  files: AnyFile<TRoute>[];
+}
+
+export interface UploadProgressEvent<TRoute extends AnyFileRoute> {
+  type: "upload-progress";
+  /**
+   * The file that is currently uploading.
+   */
+  file: UploadingFile;
+  /**
+   * All files that are being uploaded as part of this action.
+   */
+  files: AnyFile<TRoute>[];
+}
+
+export interface UploadCompletedEvent<TRoute extends AnyFileRoute> {
+  type: "upload-completed";
+  /**
+   * The file that finished uploading.
+   */
+  file: UploadedFile<TRoute>;
+  /**
+   * All files that are being uploaded as part of this action.
+   */
+  files: AnyFile<TRoute>[];
+}
+
+export interface UploadFailedEvent<TRoute extends AnyFileRoute> {
+  type: "upload-failed";
+  /**
+   * The file that failed to upload.
+   */
+  file: FailedFile<TRoute>;
+  /**
+   * All files that are being uploaded as part of this action.
+   */
+  files: AnyFile<TRoute>[];
+}
+
+export type UploadEvent<TRoute extends AnyFileRoute> =
+  | PresignedReceivedEvent<TRoute>
+  | UploadStartedEvent<TRoute>
+  | UploadProgressEvent<TRoute>
+  | UploadCompletedEvent<TRoute>
+  | UploadFailedEvent<TRoute>;
+
+export interface UploadFileOptions<TRoute extends AnyFileRoute> {
+  file: PendingFile;
+  files: AnyFile<TRoute>[];
+  input: TRoute["$types"]["input"];
+  onEvent: (event: UploadEvent<TRoute>) => void;
+
+  XHRImpl: new () => XMLHttpRequest;
+}
+
+/**
+ * Upload a file to the storage provider
+ * Throughout the upload, the file's status and progress will be updated
+ * @remarks This function never rejects
+ * @internal
```

Comment:
I haven't implemented the public-facing `uploadFiles` yet, which would take just `File[]`, transition to `PendingFile` and request the presigned URLs

Technically we could have this be public and also export a function to manually get the predesigned URLs